sleepy/types.py

```diff
@@ -613,48 +617,35 @@ def make_extract_val(self, union_ir_val: ir.Value,
-      if possible_type in narrow_to_types]
-    return UnionType(possible_types=possible_types, possible_type_nums=possible_type_nums, val_size=self.val_size)
-
-  def copy_with_extended_types(self, extended_types: List[Type],
-                               extended_type_nums: Union[List[Union[int, None]], None] = None) -> UnionType:
-    assert all(not isinstance(extended_type, UnionType) for extended_type in extended_types)
-    if extended_type_nums is None:
-      extended_type_nums = [None] * len(extended_types)
-    assert len(extended_types) == len(extended_type_nums)
-    new_possible_types = self.possible_types.copy()
-    new_possible_type_nums = self.possible_type_nums.copy()
-    for extended_type, extended_type_num in zip(extended_types, extended_type_nums):
-      if extended_type in new_possible_types:
-        continue
-      new_possible_types.append(extended_type)
-      if extended_type_num is not None and extended_type_num not in new_possible_type_nums:
-        # noinspection PyTypeChecker
-        new_possible_type_nums.append(extended_type_num)
-      else:
-        next_type_num = max(new_possible_type_nums) + 1 if len(new_possible_type_nums) > 0 else 0
-        new_possible_type_nums.append(next_type_num)
-    if self.val_size is None or any(
-      extended_type.has_unfilled_template_parameters() for extended_type in extended_types):
-      new_val_size = None
-    else:
-      new_val_size = max([self.val_size] + [extended_type.size for extended_type in extended_types])
     return UnionType(
-      possible_types=new_possible_types, possible_type_nums=new_possible_type_nums, val_size=new_val_size)
+      type_mapping=frozendict((t, self.type_mapping[t]) for t in narrow_to_types),
+      val_size=self.val_size)
 
-  @classmethod
-  def from_types(cls, possible_types: List[Type], val_size: Optional[int] = None) -> UnionType:
-    possible_types = list(dict.fromkeys(possible_types))  # possibly remove duplicates
-    possible_type_nums = list(range(len(possible_types)))
-    if val_size is None and not any(typ.has_unfilled_template_parameters() for typ in possible_types):
-      val_size = max((ctypes.sizeof(possible_type.c_type) for possible_type in possible_types), default=0)
-    return UnionType(possible_types=possible_types, possible_type_nums=possible_type_nums, val_size=val_size)
-
-  def children(self) -> List[Type]:
+  def copy_with_extended_types_and_indices(self, added_type_mapping: Mapping[Type, int]):
+    new_mapping = added_type_mapping | self.type_mapping # values from rhs are preferred
+
+    return UnionType(type_mapping=new_mapping, val_size=UnionType._new_val_size(new_mapping, self.val_size))
+
+  def copy_with_extended_types(self, extended_types: Set[Type]) -> UnionType:
+    new_mapping = dict(self.type_mapping)
+    for typ in extended_types - self.possible_types:
+      new_index = max(new_mapping.values(), default=-1) + 1
```

Comment:
i think this can also use `_first_unused_index`, doesn't matter much but maybe more consistent / compact.